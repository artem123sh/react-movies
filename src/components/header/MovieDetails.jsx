import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Logo from '../basic/Logo';
import background from './background.png';
import Poster from '../movie/Poster';
import { PRIMARY } from '../../theme';

const Title = styled.span`
  font-size: 4rem;
  opacity: 0.8;
`;

const MovieDetailsContainer = styled.header`
  flex-grow: 1;
  display: flex;
  margin-bottom: 1rem;
  padding: 5rem 5% 3rem 5%;
  position: relative;
  background-image: linear-gradient(
    rgba(0, 0, 0, 0.9), 
    rgba(0, 0, 0, 0.9)
  ), url(${background});
  background-size: 110%;
`;

const LogoContainer = styled.div`
  position: absolute;
  top: 1rem; 
  left: 5%;
  z-index: 15;
`;

const SearchIcon = styled.span`
  position: absolute;
  font-size: 3rem;
  color: ${PRIMARY};
  top: 0; 
  right: 5%;
  transform: rotate(45deg);
`;

const StyledPoster = styled(Poster)`
  top: 15%;
  left: 5%;
  flex-grow: 1;
  min-width: 20rem;
  max-width: 25rem;
`;

const Rating = styled.span`
  border: solid 1px white;
  margin: 0 2rem;
  border-radius: 50%;
  font-size: 2rem;
  min-width: 4rem;
  max-width: 4rem;
  text-align: center;
  line-height: 4rem;
  color:  ${({ voteAverage }) => {
    switch (Math.round(voteAverage / 3)) {
      case (3):
        return 'lightgreen';
      case (2):
        return 'inherit';
      case (1):
      default:
        return 'lightcoral';
    }
  }};
`;

const MovieTextDetails = styled.p`
  margin: 1.5rem 0;
  opacity: 0.8;
`;

const PrimaryMovieDetails = styled.p`
  color: ${PRIMARY};
  font-size: 2rem;
  margin: 0 3rem 0 0;
`;

const MovieDetailsRightPanel = styled.div`
  padding: 1% 5%;
  height: 100%;
  flex-grow: 1;
`;

const Section = styled.div`
  display: flex;
  align-items: center;
`;

const MovieDetails = ({
  title, releaseDate, posterPath, runtime, voteAverage, tagline, overview,
}) => (
  <MovieDetailsContainer>
    <LogoContainer><Logo /></LogoContainer>
    <Link to="/"><SearchIcon>&#9906;</SearchIcon></Link>
    <StyledPoster posterPath={posterPath} />
    <MovieDetailsRightPanel>
      <Section>
        <Title>{title}</Title>
        <Rating voteAverage={voteAverage}>{voteAverage || '0.0'}</Rating>
      </Section>
      <MovieTextDetails>{tagline}</MovieTextDetails>
      <Section>
        <PrimaryMovieDetails>{new Date(releaseDate).getFullYear()}</PrimaryMovieDetails>
        <PrimaryMovieDetails>
          {runtime}
          {' mins'}
        </PrimaryMovieDetails>
      </Section>
      <MovieTextDetails>{overview}</MovieTextDetails>
    </MovieDetailsRightPanel>
  </MovieDetailsContainer>
);

MovieDetails.propTypes = {
  title: PropTypes.string.isRequired,
  releaseDate: PropTypes.string.isRequired,
  posterPath: PropTypes.string.isRequired,
  tagline: PropTypes.string.isRequired,
  overview: PropTypes.string.isRequired,
  voteAverage: PropTypes.number.isRequired,
  runtime: PropTypes.number.isRequired,
};

export default MovieDetails;
