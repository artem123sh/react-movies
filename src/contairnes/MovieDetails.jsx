import React, { useEffect } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import MovieDetailsView from '../components/header/MovieDetails';
import { getMovie as getMovieAction } from '../store/movie/actions';

const MovieDetails = ({ getMovie, movie, error }) => {
  const { id } = useParams();
  const history = useHistory();

  useEffect(() => {
    getMovie(Number(id));
    globalThis.scrollTo(0, 0);
  }, [id]);

  useEffect(() => {
    if (error) {
      history.push('/404');
    }
  }, [error]);

  return (
    movie && (
    <MovieDetailsView
      title={movie.title}
      releaseDate={movie.release_date}
      posterPath={movie.poster_path}
      runtime={movie.runtime}
      tagline={movie.tagline}
      voteAverage={movie.vote_average}
      overview={movie.overview}
    />
    )
  );
};

MovieDetails.defaultProps = {
  error: undefined,
  movie: undefined,
};

MovieDetails.propTypes = {
  getMovie: PropTypes.func.isRequired,
  movie: PropTypes.shape(),
  error: PropTypes.shape(),
};

const mapStateToProps = (state) => {
  const { movie: { movie, error } } = state;
  return { movie, error };
};

export default connect(mapStateToProps, { getMovie: getMovieAction })(MovieDetails);
