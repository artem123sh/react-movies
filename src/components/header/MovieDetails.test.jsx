import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import MovieDetails from './MovieDetails';

describe('MovieDetails', () => {
  test('renders MovieDetails\'s snapshot', () => {
    const { asFragment } = render(
      <Router>
        <MovieDetails
          title=""
          releaseDate="2020-01-01"
          posterPath="http://404.404"
          runtime={1}
          voteAverage={1}
          tagline=""
          overview=""
        />
      </Router>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders MovieDetails\'s snapshot with default vote average', () => {
    const { asFragment } = render(
      <Router>
        <MovieDetails
          title=""
          releaseDate="2020-01-01"
          posterPath="http://404.404"
          runtime={1}
          tagline=""
          overview=""
        />
      </Router>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders MovieDetails\'s snapshot with green vote average', () => {
    const { asFragment } = render(
      <Router>
        <MovieDetails
          title=""
          releaseDate="2020-01-01"
          posterPath="http://404.404"
          runtime={1}
          voteAverage={8}
          tagline=""
          overview=""
        />
      </Router>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
