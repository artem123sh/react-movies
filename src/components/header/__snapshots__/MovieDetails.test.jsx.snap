// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails renders MovieDetails's snapshot 1`] = `
<DocumentFragment>
  <header
    class="sc-AxhCb nOGJ"
  >
    <div
      class="sc-AxhUy gvViMF"
    >
      <span
        class="sc-AxjAm hgfYjl"
      >
        netflix
      </span>
      <span
        class="sc-AxjAm cIZlLY"
      >
        roulette
      </span>
    </div>
    <a
      href="/"
    >
      <span
        class="sc-AxgMl eZZxdO"
      >
        ⚲
      </span>
    </a>
    <div
      class="sc-AxheI keyjSi"
    >
      <img
        alt="poster"
        class="sc-AxirZ jsoWpm"
        src="http://404.404"
      />
    </div>
    <div
      class="sc-fzoLsD iSxJZT"
    >
      <div
        class="sc-fzpans kOqDuY"
      >
        <span
          class="sc-AxiKw MiEfa"
        />
        <span
          class="sc-Axmtr eXOSeH"
        >
          1
        </span>
      </div>
      <p
        class="sc-AxmLO gqtsqu"
      />
      <div
        class="sc-fzpans kOqDuY"
      >
        <p
          class="sc-fzozJi fiaFkc"
        >
          2020
        </p>
        <p
          class="sc-fzozJi fiaFkc"
        >
          1 mins
        </p>
      </div>
      <p
        class="sc-AxmLO gqtsqu"
      />
    </div>
  </header>
</DocumentFragment>
`;

exports[`MovieDetails renders MovieDetails's snapshot with default vote average 1`] = `
<DocumentFragment>
  <header
    class="sc-AxhCb nOGJ"
  >
    <div
      class="sc-AxhUy gvViMF"
    >
      <span
        class="sc-AxjAm hgfYjl"
      >
        netflix
      </span>
      <span
        class="sc-AxjAm cIZlLY"
      >
        roulette
      </span>
    </div>
    <a
      href="/"
    >
      <span
        class="sc-AxgMl eZZxdO"
      >
        ⚲
      </span>
    </a>
    <div
      class="sc-AxheI keyjSi"
    >
      <img
        alt="poster"
        class="sc-AxirZ jsoWpm"
        src="http://404.404"
      />
    </div>
    <div
      class="sc-fzoLsD iSxJZT"
    >
      <div
        class="sc-fzpans kOqDuY"
      >
        <span
          class="sc-AxiKw MiEfa"
        />
        <span
          class="sc-Axmtr eXOSeH"
        >
          0.0
        </span>
      </div>
      <p
        class="sc-AxmLO gqtsqu"
      />
      <div
        class="sc-fzpans kOqDuY"
      >
        <p
          class="sc-fzozJi fiaFkc"
        >
          2020
        </p>
        <p
          class="sc-fzozJi fiaFkc"
        >
          1 mins
        </p>
      </div>
      <p
        class="sc-AxmLO gqtsqu"
      />
    </div>
  </header>
</DocumentFragment>
`;

exports[`MovieDetails renders MovieDetails's snapshot with green vote average 1`] = `
<DocumentFragment>
  <header
    class="sc-AxhCb nOGJ"
  >
    <div
      class="sc-AxhUy gvViMF"
    >
      <span
        class="sc-AxjAm hgfYjl"
      >
        netflix
      </span>
      <span
        class="sc-AxjAm cIZlLY"
      >
        roulette
      </span>
    </div>
    <a
      href="/"
    >
      <span
        class="sc-AxgMl eZZxdO"
      >
        ⚲
      </span>
    </a>
    <div
      class="sc-AxheI keyjSi"
    >
      <img
        alt="poster"
        class="sc-AxirZ jsoWpm"
        src="http://404.404"
      />
    </div>
    <div
      class="sc-fzoLsD iSxJZT"
    >
      <div
        class="sc-fzpans kOqDuY"
      >
        <span
          class="sc-AxiKw MiEfa"
        />
        <span
          class="sc-Axmtr kHYWEb"
        >
          8
        </span>
      </div>
      <p
        class="sc-AxmLO gqtsqu"
      />
      <div
        class="sc-fzpans kOqDuY"
      >
        <p
          class="sc-fzozJi fiaFkc"
        >
          2020
        </p>
        <p
          class="sc-fzozJi fiaFkc"
        >
          1 mins
        </p>
      </div>
      <p
        class="sc-AxmLO gqtsqu"
      />
    </div>
  </header>
</DocumentFragment>
`;
